packages/blocks/src/blocks.ts

```diff
@@ -1,19 +1,125 @@
-import type {
-  Block,
-  Blocks,
-  BlocksOptions,
-  Cell,
-  InferAllButLastArrayElements,
-  InferLastArrayElement,
-  InvalidCellMessage,
-} from '@/types';
 import { dump, log, raise } from '@ballerine/common';
+import type { ComponentProps, FunctionComponent, ReactNode } from 'react';
+
+/**
+ * A cell may have any property as long as it has a `type` property to discriminate a union by.
+ */
+export type Cell = { type: string } & {
+  [key: string]: unknown;
+};
+
+export type Block = Array<Cell>;
+
+export type Blocks = Array<Block>;
+
+/**
+ * Allow the consumer of `@ballerine/blocks` to register their own cell types.
+ * @example
+ * declare module '@ballerine/blocks' {
+ *    interface BlocksClient {
+ *      cells: typeof blocks; // createBlocks<TCell>();
+ *    }
+ *  }
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface BlocksClient {
+  // cells: ReturnType<typeof createBlocks<TCell>>;
+}
 
-export const createBlocks = <TCell extends Cell | InvalidCellMessage = InvalidCellMessage>(
-  options: BlocksOptions = {},
-) =>
-  new BlocksBuilder<// @ts-expect-error - A generic should always be provided.
-  TCell>([], options);
+/**
+ * Infer the generic passed into an instance of `BlocksBuilder`.
+ */
+type InferBlocksBuilderGeneric<TClass extends BlocksBuilder<any>> = TClass extends BlocksBuilder<
+  infer TCell
+>
+  ? TCell
+  : never;
+
+/**
+ * Infer the generic passed into the instance of `BlocksBuilder` registered by the consumer of `@ballerine/blocks`.
+ */
+export type Cells =
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  InferBlocksBuilderGeneric<BlocksClient['cells']>;
+
+/**
+ * Extract the type of cell from the registered cells.
+ */
+export type CellType = Cells['type'];
+
+/**
+ * Extract the props of a cell by type from the registered cells.
+ * @example
+ * const Cell: FunctionComponent<ExtractCellProps<'text'>> = ({ text }) => <p>{text}</p>;
+ */
+export type ExtractCellProps<TType extends CellType> = Omit<
+  Extract<Cells, { type: TType }>,
+  'type'
+>;
+
+/**
+ * A type to help build objects of cell components.
+ * @example
+ * const cells: CellsMap = {
+ *    text: TextCell,
+ *    image: ImageCell,
+ *  }
+ */
+export type CellsMap = {
+  [TType in CellType]: FunctionComponent<ExtractCellProps<TType>>;
+};
+
+export type InferAllButLastArrayElements<T extends Array<any>> = T extends [...infer U, any]
+  ? U
+  : [];
+
+export type InferLastArrayElement<T extends Array<any>> = T extends [...any, infer U] ? U : never;
```

Comment:
Maybe `get` is more intuitive than `infer`? Your choice :)
```suggestion
export type GetLastArrayElement<T extends Array<any>> = T extends [...any, infer U] ? U : never;
```

Reply:
Following existing conventions from libraries I know like `tRPC` and Next.JS where `infer` is preferred.